File: lintcode/easy/604_window_sum.py
'''
Link: https://www.lintcode.com/problem/window-sum/description

Relevant problems:
http://www.lintcode.com/problem/sliding-window-median/
http://www.lintcode.com/problem/sliding-window-maximum/
'''

# My own solution.
class Solution:
    """
    @param nums: a list of integers.
    @param k: length of window.
    @return: the sum of the element inside the window at each moving.
    """
    def winSum(self, nums, k):
        # write your code here
        n = len(nums)
        if n <= 0:
            return []
        if n <= k:
            return [sum(nums)]
        res = []
        sum_val = sum(nums[:k])
        res.append(sum_val)
        for i in range(n - k):
            sum_val = sum_val - nums[i] + nums[i + k]
            res.append(sum_val)
        return res

# A faster solution. The official one from Jiuzhang.com.    
class Solution:
    """
    @param nums: a list of integers.
    @param k: length of window.
    @return: the sum of the element inside the window at each moving.
    """
    def winSum(self, nums, k):
        # write your code here
        n = len(nums)
        if n < k or k <= 0:
            return []
        res = [0] * (n - k + 1)
        res[0] = sum(nums[:k])
        for i in range(1, n - k + 1):
            res[i] = res[i - 1] - nums[i - 1] + nums[i + k - 1]
            
        return res

File: lintcode/easy/604_window_sum.java
public class Solution {
    /**
     * @param nums: a list of integers.
     * @param k: length of window.
     * @return: the sum of the element inside the window at each moving.
     */
    public int[] winSum(int[] nums, int k) {
        // write your code here
        int n = nums.length;
        if (n <= 0){
            return new int [0];
        }
        if (n <= k){
            int[] res = new int [1];
            res[0] = Arrays.stream(nums).sum();
            return res;
        }
        int sum_val = Arrays.stream(nums, 0, k).sum();
        int[] res = new int [n - k + 1];
        res[0] = Arrays.stream(nums, 0, k).sum();
        for (int i = 0; i < res.length - 1; i++){
            res[i + 1] = res[i] - nums[i] + nums[i + k];
        }
        return res;
    }
}
